source/grid/grid_in.cc

```diff
@@ -2184,483 +2217,495 @@ GridIn<dim, spacedim>::read_msh(std::istream &in)
-  // 1 or codim 2.
-  std::map<unsigned int, unsigned int> vertex_counts;
+          // Assert we reached the end of the block
+          in >> line;
+          static const std::string end_nodes_marker[] = {"$ENDNOD",
+                                                         "$EndNodes"};
+          AssertThrow(line == end_nodes_marker[gmsh_file_format == 10 ? 0 : 1],
+                      ExcInvalidGMSHInput(line));
+        }
 
-  {
-    static constexpr std::array<unsigned int, 8> local_vertex_numbering = {
-      {0, 1, 5, 4, 2, 3, 7, 6}};
-    unsigned int global_cell = 0;
-    for (int entity_block = 0; entity_block < n_entity_blocks; ++entity_block)
-      {
-        unsigned int  material_id;
-        unsigned long numElements;
-        int           cell_type;
+      // if the next block is of kind $Elements or $ELM, parse it
+      if (line == begin_elements_marker[gmsh_file_format == 10 ? 0 : 1])
+        {
+          // now read the cell list
+          if (gmsh_file_format > 40)
+            {
+              int min_node_tag;
+              int max_node_tag;
+              in >> n_entity_blocks >> n_cells >> min_node_tag >> max_node_tag;
+            }
+          else if (gmsh_file_format == 40)
+            {
+              in >> n_entity_blocks >> n_cells;
+            }
+          else
+            {
+              n_entity_blocks = 1;
+              in >> n_cells;
+            }
 
-        if (gmsh_file_format < 40)
-          {
-            material_id = 0;
-            cell_type   = 0;
-            numElements = n_cells;
-          }
-        else if (gmsh_file_format == 40)
-          {
-            int tagEntity, dimEntity;
-            in >> tagEntity >> dimEntity >> cell_type >> numElements;
-            material_id = tag_maps[dimEntity][tagEntity];
-          }
-        else
-          {
-            // for gmsh_file_format 4.1 the order of tag and dim is reversed,
-            int tagEntity, dimEntity;
-            in >> dimEntity >> tagEntity >> cell_type >> numElements;
-            material_id = tag_maps[dimEntity][tagEntity];
-          }
+          // // set up array of cells and subcells (faces). In 1d, there is
+          // currently no
+          // // standard way in deal.II to pass boundary indicators attached to
+          // // individual vertices, so do this by hand via the boundary_ids_1d
+          // array std::vector<CellData<dim>>                 cells; SubCellData
+          // subcelldata; std::map<unsigned int, types::boundary_id>
+          // boundary_ids_1d;
+
+          // // Track the number of times each vertex is used in 1D. This
+          // determines
+          // // whether or not we can assign a boundary id to a vertex. This is
+          // necessary
+          // // because sometimes gmsh saves internal vertices in the $ELEM list
+          // in codim
+          // // 1 or codim 2.
+          // std::map<unsigned int, unsigned int> vertex_counts;
```

Comment:
This commented block of code is not necessary anymore, you have moved that elsewhere above